Sources/AWSSDKSwiftCore/MetaDataService.swift

```diff
@@ -180,32 +184,48 @@ struct InstanceMetaDataServiceProvider: MetaDataServiceProvider {
 
     typealias MetaData = InstanceMetaData
 
+    static let instanceMetadataApiTokenUri = "/latest/api/token"
     static let instanceMetadataUri = "/latest/meta-data/iam/security-credentials/"
     static var host = "169.254.169.254"
+    static var apiTokenURL: String {
+        return "http://\(host)\(instanceMetadataApiTokenUri)"
+    }
     static var baseURLString: String {
         return "http://\(host)\(instanceMetadataUri)"
     }
 
-    func uri(eventLoopGroup: EventLoopGroup) -> EventLoopFuture<String> {
+    func getCredential(eventLoopGroup: EventLoopGroup) -> EventLoopFuture<CredentialProvider> {
+        //  no point storing the session key as the credentials last as long
+        var sessionTokenHeader: [String: String] = [:]
         // instance service expects absoluteString as uri...
-        return request(uri:InstanceMetaDataServiceProvider.baseURLString, timeout: 2, eventLoopGroup: eventLoopGroup)
-            .flatMapThrowing{ response in
-                switch response.head.status {
-                case .ok:
-                    let roleName = String(data: response.body, encoding: .utf8) ?? ""
-                    return "\(InstanceMetaDataServiceProvider.baseURLString)/\(roleName)"
-                default:
-                    throw MetaDataServiceError.couldNotGetInstanceRoleName
+        return request(uri:InstanceMetaDataServiceProvider.apiTokenURL, method: .PUT, headers:["X-aws-ec2-metadata-token-ttl-seconds":"21600"], timeout: 2, eventLoopGroup: eventLoopGroup).flatMapThrowing { response in
```

Comment:
line breaks? maybe? please ;)

what do you think about return an optional dict of headers?

Reply:
The sessionTokenHeaders are used in multiple places (I also need them when getting credentials) within the process. I could pass them down through all the flatMaps but thats quite awkward so just leaving them at the top and using variable capture seemed easier